### ISSUE TITLE
Type inference degradation with class methods that take function arguments

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#176]
This module compiles without issue:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DefaultSignatures #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

import Data.Kind (Type)
import Data.Singletons.Prelude
import Data.Singletons.TH

$(singletons [d|
  class Foo1 a where
    bar1 :: a -> (a -> b) -> b
    baz1 :: a

  {-
  quux1 :: Foo1 a => a -> a
  quux1 x = x `bar1` \_ -> baz1
  -}

  class Foo2 a where
    bar2 :: a -> b -> b
    baz2 :: a

  quux2 :: Foo2 a => a -> a
  quux2 x = x `bar2` baz2
  |])
```

But if you uncomment out the definition of `quux1`:

```
Bug.hs:16:3: error:
    • Could not deduce (SFoo1 k3) arising from a use of ‘sBaz1’
      from the context: SFoo1 a0
        bound by the type signature for:
                   sQuux1 :: SFoo1 a0 => Sing t0 -> Sing (Apply Quux1Sym0 t0)
        at Bug.hs:(16,3)-(30,4)
      or from: t0 ~ x0
        bound by the type signature for:
                   lambda_a1G3Q :: t0 ~ x0 => Sing x0 -> Sing (Apply Quux1Sym0 t0)
        at Bug.hs:(16,3)-(30,4)
      or from: _z_16277899320 ~ arg_16277899150
        bound by the type signature for:
                   lambda_a1G3U :: _z_16277899320 ~ arg_16277899150 =>
                                   Sing _z_16277899320
                                   -> Sing (Case_1627789930 x0 arg_16277899150 _z_16277899320)
        at Bug.hs:(16,3)-(30,4)
      Possible fix:
        add (SFoo1 k3) to the context of
          the type signature for:
            lambda_a1G3U :: _z_16277899320 ~ arg_16277899150 =>
                            Sing _z_16277899320
                            -> Sing (Case_1627789930 x0 arg_16277899150 _z_16277899320)
    • In the expression: sBaz1
      In an equation for ‘lambda’: lambda _z_1627789932_a1G3V = sBaz1
      In the expression:
        let
          lambda_a1G3U ::
            forall _z_1627789932_a1G3k.
            _z_1627789932_a1G3k ~ arg_1627789915_a1G3f =>
            Sing _z_1627789932_a1G3k
            -> Sing (Case_1627789930 x_a1G3c arg_1627789915_a1G3f _z_1627789932_a1G3k)
          lambda_a1G3U _z_1627789932_a1G3V = sBaz1
        in lambda_a1G3U _s_z_1627789932
```

I suspect that the fact that `bar1` accepts a function as its second argument has something to do with it, since `quux2` typechecks just fine.

The issue even persists if you define `quux1` in terms of a concrete `Foo1` instance:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DefaultSignatures #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

import Data.Kind (Type)
import Data.Singletons.Prelude
import Data.Singletons.TH

$(singletons [d|
  class Foo1 a where
    bar1 :: a -> (a -> b) -> b
    baz1 :: a

  instance Foo1 Bool where
    bar1 x f = f x
    baz1     = True

  quux1 :: Bool -> Bool
  quux1 x = x `bar1` \_ -> baz1
  |])
```

```
Bug.hs:17:3: error:
    • Could not deduce (SFoo1 k3) arising from a use of ‘sBaz1’
      from the context: t0 ~ x0
        bound by the type signature for:
                   lambda_a1GuD :: t0 ~ x0 => Sing x0 -> Sing (Apply Quux1Sym0 t0)
        at Bug.hs:(17,3)-(28,4)
      or from: _z_16277915810 ~ arg_16277915690
        bound by the type signature for:
                   lambda_a1GuH :: _z_16277915810 ~ arg_16277915690 =>
                                   Sing _z_16277915810
                                   -> Sing (Case_1627791579 x0 arg_16277915690 _z_16277915810)
        at Bug.hs:(17,3)-(28,4)
      Possible fix:
        add (SFoo1 k3) to the context of
          the type signature for:
            lambda_a1GuH :: _z_16277915810 ~ arg_16277915690 =>
                            Sing _z_16277915810
                            -> Sing (Case_1627791579 x0 arg_16277915690 _z_16277915810)
    • In the expression: sBaz1
      In an equation for ‘lambda’: lambda _z_1627791581_a1GuI = sBaz1
      In the expression:
        let
          lambda_a1GuH ::
            forall _z_1627791581_a1GtV.
            _z_1627791581_a1GtV ~ arg_1627791569_a1GtQ =>
            Sing _z_1627791581_a1GtV
            -> Sing (Case_1627791579 x_a1GtN arg_1627791569_a1GtQ _z_1627791581_a1GtV)
          lambda_a1GuH _z_1627791581_a1GuI = sBaz1
        in lambda_a1GuH _s_z_1627791581
```

**Comment by RyanGlScott** [comment]
Like #175, this might have resulted from GHC being unable to infer kinds in some places. To illustrate, if you dump all of the generated code for the `Bool` example above into a file:

```haskell
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DefaultSignatures #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug2 where

import Data.Kind (Type)
import Data.Singletons.Prelude
import Data.Singletons.SuppressUnusedWarnings

type family Case_1627771676_a1BiR x_a1BiM
                                  arg_1627771666_a1BiP
                                  t_a1BiV where
  Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP _z_1627771678_a1BiU = Baz1Sym0
type family Lambda_1627771673_a1BiO x_a1BiM t_a1BiW where
  Lambda_1627771673_a1BiO x_a1BiM arg_1627771666_a1BiP = Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP arg_1627771666_a1BiP
type Lambda_1627771673Sym2 t_a1BiX t_a1BiY =
    Lambda_1627771673_a1BiO t_a1BiX t_a1BiY
instance SuppressUnusedWarnings Lambda_1627771673Sym1 where
  suppressUnusedWarnings _
    = snd
        ((,) Lambda_1627771673Sym1KindInference ())
data Lambda_1627771673Sym1 l_a1Bj2 l_a1Bj1
  = forall arg_a1Bj3. SameKind (Apply (Lambda_1627771673Sym1 l_a1Bj2) arg_a1Bj3) (Lambda_1627771673Sym2 l_a1Bj2 arg_a1Bj3) =>
    Lambda_1627771673Sym1KindInference
type instance Apply (Lambda_1627771673Sym1 l_a1Bj2) l_a1Bj1 = Lambda_1627771673Sym2 l_a1Bj2 l_a1Bj1
instance SuppressUnusedWarnings Lambda_1627771673Sym0 where
  suppressUnusedWarnings _
    = snd
        ((,) Lambda_1627771673Sym0KindInference ())
data Lambda_1627771673Sym0 l_a1BiZ
  = forall arg_a1Bj0. SameKind (Apply Lambda_1627771673Sym0 arg_a1Bj0) (Lambda_1627771673Sym1 arg_a1Bj0) =>
    Lambda_1627771673Sym0KindInference
type instance Apply Lambda_1627771673Sym0 l_a1BiZ = Lambda_1627771673Sym1 l_a1BiZ
type Quux1Sym1 (t_a1BiI :: Bool) = Quux1 t_a1BiI
instance SuppressUnusedWarnings Quux1Sym0 where
  suppressUnusedWarnings _
    = snd ((,) Quux1Sym0KindInference ())
data Quux1Sym0 (l_a1BiJ :: TyFun Bool Bool)
  = forall arg_a1BiK. SameKind (Apply Quux1Sym0 arg_a1BiK) (Quux1Sym1 arg_a1BiK) =>
    Quux1Sym0KindInference
type instance Apply Quux1Sym0 l_a1BiJ = Quux1Sym1 l_a1BiJ
type family Quux1 (a_a1BiL :: Bool) :: Bool where
  Quux1 x_a1BiM = Apply (Apply Bar1Sym0 x_a1BiM) (Apply Lambda_1627771673Sym0 x_a1BiM)
type Bar1Sym2 (t_a1Bj6 :: a1627771661)
              (t_a1Bj7 :: TyFun a1627771661 b1627771662 -> Type) =
    Bar1 t_a1Bj6 t_a1Bj7
instance SuppressUnusedWarnings Bar1Sym1 where
  suppressUnusedWarnings _
    = snd ((,) Bar1Sym1KindInference ())
data Bar1Sym1 (l_a1Bjb :: a1627771661)
              (l_a1Bja :: TyFun (TyFun a1627771661 b1627771662
                                 -> Type) b1627771662)
  = forall arg_a1Bjc. SameKind (Apply (Bar1Sym1 l_a1Bjb) arg_a1Bjc) (Bar1Sym2 l_a1Bjb arg_a1Bjc) =>
    Bar1Sym1KindInference
type instance Apply (Bar1Sym1 l_a1Bjb) l_a1Bja = Bar1Sym2 l_a1Bjb l_a1Bja
instance SuppressUnusedWarnings Bar1Sym0 where
  suppressUnusedWarnings _
    = snd ((,) Bar1Sym0KindInference ())
data Bar1Sym0 (l_a1Bj8 :: TyFun a1627771661 (TyFun (TyFun a1627771661 b1627771662
                                                    -> Type) b1627771662
                                             -> Type))
  = forall arg_a1Bj9. SameKind (Apply Bar1Sym0 arg_a1Bj9) (Bar1Sym1 arg_a1Bj9) =>
    Bar1Sym0KindInference
type instance Apply Bar1Sym0 l_a1Bj8 = Bar1Sym1 l_a1Bj8
type Baz1Sym0 = Baz1
class PFoo1 (a_a1BiB :: Type) where
  type Bar1 (arg_a1Bj4 :: a_a1BiB)
            (arg_a1Bj5 :: TyFun a_a1BiB b_a1BiC -> Type) :: b_a1BiC
  type Baz1 :: a_a1BiB
type family Bar1_1627771712_a1Bjr (a_a1Bjo :: Bool)
                                  (a_a1Bjp :: TyFun Bool b_a1BiC -> Type) :: b_a1BiC where
  Bar1_1627771712_a1Bjr x_a1Bjm f_a1Bjn = Apply f_a1Bjn x_a1Bjm
type Bar1_1627771712Sym2 (t_a1Bjs :: Bool)
                         (t_a1Bjt :: TyFun Bool b1627771662 -> Type) =
    Bar1_1627771712_a1Bjr t_a1Bjs t_a1Bjt
instance SuppressUnusedWarnings Bar1_1627771712Sym1 where
  suppressUnusedWarnings _
    = snd ((,) Bar1_1627771712Sym1KindInference ())
data Bar1_1627771712Sym1 (l_a1Bjx :: Bool)
                         (l_a1Bjw :: TyFun (TyFun Bool b1627771662 -> Type) b1627771662)
  = forall arg_a1Bjy. SameKind (Apply (Bar1_1627771712Sym1 l_a1Bjx) arg_a1Bjy) (Bar1_1627771712Sym2 l_a1Bjx arg_a1Bjy) =>
    Bar1_1627771712Sym1KindInference
type instance Apply (Bar1_1627771712Sym1 l_a1Bjx) l_a1Bjw = Bar1_1627771712Sym2 l_a1Bjx l_a1Bjw
instance SuppressUnusedWarnings Bar1_1627771712Sym0 where
  suppressUnusedWarnings _
    = snd ((,) Bar1_1627771712Sym0KindInference ())
data Bar1_1627771712Sym0 (l_a1Bju :: TyFun Bool (TyFun (TyFun Bool b1627771662
                                                        -> Type) b1627771662
                                                 -> Type))
  = forall arg_a1Bjv. SameKind (Apply Bar1_1627771712Sym0 arg_a1Bjv) (Bar1_1627771712Sym1 arg_a1Bjv) =>
    Bar1_1627771712Sym0KindInference
type instance Apply Bar1_1627771712Sym0 l_a1Bju = Bar1_1627771712Sym1 l_a1Bju
type family Baz1_1627771721_a1BjA :: Bool where
  Baz1_1627771721_a1BjA = TrueSym0
type Baz1_1627771721Sym0 = Baz1_1627771721_a1BjA
instance PFoo1 Bool where
  type Bar1 (a_a1Bjo :: Bool) (a_a1Bjp :: TyFun Bool b_a1BiC
                                          -> Type) = Apply (Apply Bar1_1627771712Sym0 a_a1Bjo) a_a1Bjp
  type Baz1 = Baz1_1627771721Sym0
class SFoo1 a_a1BiB where
  sBar1 ::
    forall (t_a1BjI :: a_a1BiB)
           (t_a1BjJ :: TyFun a_a1BiB b_a1BiC -> Type).
    Sing t_a1BjI
    -> Sing t_a1BjJ
       -> Sing (Apply (Apply Bar1Sym0 t_a1BjI) t_a1BjJ :: b_a1BiC)
  sBaz1 :: Sing (Baz1Sym0 :: a_a1BiB)
instance SFoo1 Bool where
  sBar1 ::
    forall (t_a1BjK :: Bool) (t_a1BjL :: TyFun Bool b_a1BiC -> Type).
    Sing t_a1BjK
    -> Sing t_a1BjL
       -> Sing (Apply (Apply Bar1Sym0 t_a1BjK) t_a1BjL :: b_a1BiC)
  sBaz1 :: Sing (Baz1Sym0 :: Bool)
  sBar1 sX sF
    = let
        lambda_a1BjM ::
          forall x_a1Bjm f_a1Bjn.
          (t_a1BjK ~ x_a1Bjm, t_a1BjL ~ f_a1Bjn) =>
          Sing x_a1Bjm
          -> Sing f_a1Bjn
             -> Sing (Apply (Apply Bar1Sym0 t_a1BjK) t_a1BjL :: b_a1BiC)
        lambda_a1BjM x_a1BjN f_a1BjO = applySing f_a1BjO x_a1BjN
      in lambda_a1BjM sX sF
  sBaz1 = STrue

-------------------------------------------------------------------------------

sQuux1 ::
  forall (t_a1BjB :: Bool).
  Sing t_a1BjB -> Sing (Apply Quux1Sym0 t_a1BjB :: Bool)
sQuux1 sX
  = let
      lambda_a1BjC ::
        forall x_a1BiM.
        t_a1BjB ~ x_a1BiM =>
        Sing x_a1BiM -> Sing (Apply Quux1Sym0 t_a1BjB :: Bool)
      lambda_a1BjC x_a1BjD
        = applySing
            (applySing (singFun2 (Proxy :: Proxy Bar1Sym0) sBar1) x_a1BjD)
            (singFun1
               (Proxy :: Proxy (Apply Lambda_1627771673Sym0 x_a1BiM))
               (\ sArg_1627771666
                  -> let
                       lambda_a1BjE ::
                         forall arg_1627771666_a1BiP.
                         Sing arg_1627771666_a1BiP
                         -> Sing (Apply (Apply Lambda_1627771673Sym0 x_a1BiM) arg_1627771666_a1BiP)
                       lambda_a1BjE arg_1627771666_a1BjF
                         = case arg_1627771666_a1BjF of {
                             _s_z_1627771678
                               -> let
                                    lambda_a1BjG ::
                                      forall _z_1627771678_a1BiU.
                                      _z_1627771678_a1BiU ~ arg_1627771666_a1BiP =>
                                      Sing _z_1627771678_a1BiU
                                      -> Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP _z_1627771678_a1BiU)
                                    lambda_a1BjG _z_1627771678_a1BjH = sBaz1
                                  in lambda_a1BjG _s_z_1627771678 } ::
                             Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP arg_1627771666_a1BiP)
                     in lambda_a1BjE sArg_1627771666))
    in lambda_a1BjC sX
```

Then I was able to make it typecheck by adding some kind annotations in `lambda_a1BjE` (located within `sQuux1`):

```diff
154c154
<                          -> Sing (Apply (Apply Lambda_1627771673Sym0 x_a1BiM) arg_1627771666_a1BiP)
---
>                          -> Sing (Apply (Apply Lambda_1627771673Sym0 x_a1BiM) arg_1627771666_a1BiP :: Bool)
163c163
<                                       -> Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP _z_1627771678_a1BiU)
---
>                                       -> Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP _z_1627771678_a1BiU :: Bool)
166c166
<                              Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP arg_1627771666_a1BiP)
---
>                              Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP arg_1627771666_a1BiP :: Bool)
```

**Comment by goldfirere** [comment]
I think this is a manifestation of the second infelicity listed at the end of the README. If I recall, this is not an easy problem to solve. Note that your kind annotation `:: Bool` is not obvious at all, as `quux1` mentions `Bool` nowhere. One of my guiding principles in implementing `singletons` is that we never ever ever do type inference. Not even a little bit. Because once you start, you can't stop. Coming up with the fact that you need `Bool` in a kind annotation looks like kind inference to me.

So you see a better way?

Thus, while I admit that this is indeed a bug, I don't see a way forward here.

**Comment by RyanGlScott** [comment]
Indeed, I shudder to think of how much effort it would take to reinvent GHC's type inferencer in Template Haskell. My hope is that perhaps this bug could be worked around by providing explicit type ascriptions to the right terms or patterns. However, my brief attempts at this were stymied. When I tried adding a type ascription to a term:

```haskell

  quux1 :: Bool -> Bool
  quux1 x = x `bar1` \_ -> (baz1 :: Bool)
```

`singletons` didn't like that:

```
Bug.hs:17:3: error:                                                                       
    Singling of explicit type annotations not yet supported.
```

And when I tried adding a type ascription to a pattern, I was reminded of this wonderful limitation of Template Haskell on GHC 8.0:

```
Bug.hs:27:22: error:
    Type signatures in patterns not (yet) handled by Template Haskell
      _ :: Bool
```

This might be surmountable if I tried building `singletons` on GHC HEAD against a version of `th-desugar` that incorporated https://github.com/goldfirere/th-desugar/pull/48, though.

**Comment by goldfirere** [comment]
Ah. If you wish to allow the user to provide the type ascription, then perhaps `singletons` can evolve to handle singling them. I will meditate on this possibility.

**Comment by RyanGlScott** [comment]
It's one possible avenue... albeit a somewhat unsatisfying one. Let me be upfront and state that the reason I want the ability to do this is because I want to implement `Monad` (i.e., https://github.com/goldfirere/singletons/issues/37), which requires:

```haskell
$(singletonsOnly [d|
  class Monad m where
    (>>=) :: m a -> (a -> m b) -> m b
    (>>) :: m a -> m b -> m b
    m >> k = m >>= \_ -> k
  |])
```

You could rewrite `m >>= \_ -> k` to avoid lambdas. But for the same reason, you'd also be unable to use `do`-notation or list comprehensions without completely rewriting them to a different form.

-----

FWIW, it turns out you don't need to infer `Bool` to make `sQuux1` typecheck—you can get away with just an `SFoo1` constraint:

```haskell
sQuux1 ::
  forall (t_a1BjB :: Bool).
  Sing t_a1BjB -> Sing (Apply Quux1Sym0 t_a1BjB :: Bool)
sQuux1 sX
  = let
      lambda_a1BjC ::
        forall x_a1BiM.
        t_a1BjB ~ x_a1BiM =>
        Sing x_a1BiM -> Sing (Apply Quux1Sym0 t_a1BjB :: Bool)
      lambda_a1BjC x_a1BjD
        = applySing
            (applySing (singFun2 (Proxy :: Proxy Bar1Sym0) sBar1) x_a1BjD)
            (singFun1
               (Proxy :: Proxy (Apply Lambda_1627771673Sym0 x_a1BiM))
               (\ sArg_1627771666
                  -> let
                       lambda_a1BjE ::
                         forall arg_1627771666_a1BiP ki.
                         SFoo1 ki =>
                         Sing arg_1627771666_a1BiP
                         -> Sing (Apply (Apply Lambda_1627771673Sym0 x_a1BiM) arg_1627771666_a1BiP :: ki)
                       lambda_a1BjE arg_1627771666_a1BjF
                         = case arg_1627771666_a1BjF of {
                             _s_z_1627771678
                               -> let
                                    lambda_a1BjG ::
                                      forall _z_1627771678_a1BiU.
                                      _z_1627771678_a1BiU ~ arg_1627771666_a1BiP =>
                                      Sing _z_1627771678_a1BiU
                                      -> Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP _z_1627771678_a1BiU :: ki)
                                    lambda_a1BjG _z_1627771678_a1BjH = sBaz1
                                  in lambda_a1BjG _s_z_1627771678 } ::
                             Sing (Case_1627771676_a1BiR x_a1BiM arg_1627771666_a1BiP arg_1627771666_a1BiP :: ki)
                     in lambda_a1BjE sArg_1627771666))
    in lambda_a1BjC sX
```

Note the additional `ki` type variable (which also has an `SFoo1` constraint) in `lambda_a1BjE`. GHC can at least infer that `ki ~ Bool`, which is better than before.

**Comment by goldfirere** [comment]
But how can we know to use this `SFoo1` constraint?

Your motivating example is helpful, and indeed motivating. And you're right that a user annotation won't really help us here in the long run. But I'm not sure how else to proceed.

Wait a tick. The reason I do all this `let`-binding of lambdas is only for the scoped type variables. But now that TH supports pattern signatures, maybe we can use the pattern signatures to bind the scoped type variables. This might simplify a great deal, as that `let`-binding stuff is a major pain.

This would be a non-trivial amount of work, though. I doubt this end of the operation is well-documented enough for someone else to make the change. What's your real end-goal here? Do you have an application in mind to get me excited about?

**Comment by RyanGlScott** [comment]
> Do you have an application in mind to get me excited about?

I don't—this is the result of me trying to figure out how much work it would take to fix #37 by just singling the unaltered definition of `Monad` and friends. Luckily, this bug seems to be the last major blocker, and it's possible to work around if you're willing to shuffle around some definitions to avoid lambdas in some places, e.g.,

```haskell
(>>) :: Monad m => m a -> m b -> m b
m >> k = m >>= const k

liftM :: Monad m => (a1 -> r) -> m a1 -> m r
liftM f m1 = m1 >>= return . f

liftM2  :: Monad m => (a1 -> a2 -> r) -> m a1 -> m a2 -> m r
liftM2 f m1 m2 = liftM f m1 >>= (`liftM` m2)

-- etc.
```

In fact, I now have working promoted and singled versions of `Functor`, `Applicative`, and `Monad`, so perhaps that's something that could be worked into `singletons` at some point...

Ultimately, the biggest gain out of fixing this bug would be additional syntactic niceties (`do`-notation and list comprehensions), and if you feel fixing this bug would require heroic implementation efforts, then it's probably not worth the ROI for now.
